base/dates/rounding.jl

```diff
@@ -160,21 +231,54 @@ function Base.round(dt::TimeType, p::Period, r::RoundingMode{:NearestTiesUp})
     return (dt - f) < (c - dt) ? f : c
 end
 
-Base.round(dt::TimeType, p::Period, r::RoundingMode{:Down}) = Base.floor(dt, p)
-Base.round(dt::TimeType, p::Period, r::RoundingMode{:Up}) = Base.ceil(dt, p)
+"""
+    round(x::Period, precision::T, [r::RoundingMode]) where T <: Union{TimePeriod, Week, Day} -> T
+
+Rounds `x` to the nearest multiple of `precision`. If `x` and `precision` are different
+subtypes of `Period`, the return value will have the same type as `precision`. By default
+(`RoundNearestTiesUp`), ties (e.g., rounding 90 minutes to the nearest hour) will be rounded
+up.
+
+For convenience, `precision` may be a type instead of a value: `round(x, Dates.Hour)` is a
+shortcut for `round(x, Dates.Hour(1))`.
+
+```jldoctest
+julia> round(Dates.Day(16), Dates.Week)
+2 weeks
+
+julia> round(Dates.Minute(44), Dates.Minute(15))
+45 minutes
+
+julia> round(Dates.Hour(36), Dates.Day)
+3 days
+```
+
+Valid rounding modes for `round(::Period, ::T, ::RoundingMode)` are `RoundNearestTiesUp`
+(default), `RoundDown` (`floor`), and `RoundUp` (`ceil`).
+
+Rounding to a `precision` of `Month`s or `Year`s is not supported, as these `Period`s are of
+inconsistent length.
+"""
+function Base.round(x::ConvertiblePeriod, precision::ConvertiblePeriod, r::RoundingMode{:NearestTiesUp})
+    f, c = floorceil(x, precision)
+    return (Nanosecond(x) - Nanosecond(f)) < (Nanosecond(c) - Nanosecond(x)) ? f : c
+end
+
+Base.round(x::TimeTypePeriod, p::Period, r::RoundingMode{:Down}) = Base.floor(x, p)
+Base.round(x::TimeTypePeriod, p::Period, r::RoundingMode{:Up}) = Base.ceil(x, p)
 
 # No implementation of other `RoundingMode`s: rounding to nearest "even" is skipped because
 # "even" is not defined for Period; rounding toward/away from zero is skipped because ISO
 # 8601's year 0000 is not really "zero".
-Base.round(::TimeType, p::Period, ::RoundingMode) = throw(DomainError(p))
+Base.round(::TimeTypePeriod, p::Period, ::RoundingMode) = throw(DomainError(p))
 
 # Default to RoundNearestTiesUp.
-Base.round(dt::TimeType, p::Period) = Base.round(dt, p, RoundNearestTiesUp)
+Base.round(x::TimeTypePeriod, p::Period) = Base.round(x, p, RoundNearestTiesUp)
 
 # Make rounding functions callable using Period types in addition to values.
-Base.floor(dt::TimeType, p::Type{<:Period}) = Base.floor(dt, p(1))
-Base.ceil(dt::TimeType, p::Type{<:Period}) = Base.ceil(dt, p(1))
+Base.floor(x::TimeTypePeriod, p::Type{<:Period}) = Base.floor(x, p(1))
```

Comment:
Should probably be changed to:
```julia
Base.floor(x::TimeTypePeriod, ::Type{P}) where P <: Period = Base.floor(x, oneunit(P))
```

Reply:
Interesting. I was completely unaware of `oneunit`.

Reply:
It's the unit aware sibling to `one`